logger.go

```diff
@@ -296,12 +297,22 @@ func (log *Logger) check(lvl zapcore.Level, msg string) *zapcore.CheckedEntry {
 
 	// Thread the error output through to the CheckedEntry.
 	ce.ErrorOutput = log.errorOutput
-	if log.addCaller {
-		ce.Entry.Caller = zapcore.NewEntryCaller(runtime.Caller(log.callerSkip + callerSkipOffset))
-		if !ce.Entry.Caller.Defined {
+	if log.addCaller || log.addFunction {
+		pc := make([]uintptr, 1)
+		numFrames := runtime.Callers(log.callerSkip+callerSkipOffset+1, pc)
+		frame, _ := runtime.CallersFrames(pc[:numFrames]).Next()
```

Comment:
If I've done the benchmark correctly, there should be no additional performance impact. 

I ran benchmark on macbook pro with:
- OS version 10.13.6
- Memory 16 GB 2400 MHz DDR4
- Processor 2,6 GHz Intel Core i7

I ran the benchmarks on Zap.Check separately to avoid cpu thermal scaling. 

### Before changes (master branch)
```
➜  benchmarks git:(master) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkDisabledWithoutFields/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkDisabledWithoutFields/Zap.Check-12             1000000000             1.12 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    1.382s
➜  benchmarks git:(master) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkDisabledAccumulatedContext/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkDisabledAccumulatedContext/Zap.Check-12             1000000000             1.15 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    1.470s
➜  benchmarks git:(master) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkDisabledAddingFields/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkDisabledAddingFields/Zap.Check-12             1000000000             1.23 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    1.574s
➜  benchmarks git:(master) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkWithoutFields/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkWithoutFields/Zap.Check-12              9375018           133 ns/op           0 B/op           0 allocs/op
BenchmarkWithoutFields/Zap.CheckSampled-12      30395667            37.7 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    2.726s
➜  benchmarks git:(master) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkAccumulatedContext/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkAccumulatedContext/Zap.Check-12              9057684           136 ns/op           0 B/op           0 allocs/op
BenchmarkAccumulatedContext/Zap.CheckSampled-12      27883348            40.9 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    2.726s
➜  benchmarks git:(master) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkAddingFields/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkAddingFields/Zap.Check-12              1000000          1159 ns/op         771 B/op           5 allocs/op
BenchmarkAddingFields/Zap.CheckSampled-12       7131936           167 ns/op          87 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    2.681s
```

### After function name changes
```
➜  benchmarks git:(function-name) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkDisabledWithoutFields/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkDisabledWithoutFields/Zap.Check-12         	1000000000	         1.12 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	go.uber.org/zap/benchmarks	1.369s
➜  benchmarks git:(function-name) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkDisabledAccumulatedContext/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkDisabledAccumulatedContext/Zap.Check-12             1000000000             1.13 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    1.355s
➜  benchmarks git:(function-name) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkDisabledAddingFields/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkDisabledAddingFields/Zap.Check-12             1000000000             1.29 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    1.596s
➜  benchmarks git:(function-name) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkWithoutFields/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkWithoutFields/Zap.Check-12             11146232           121 ns/op           0 B/op           0 allocs/op
BenchmarkWithoutFields/Zap.CheckSampled-12      31141713            37.3 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    2.872s
➜  benchmarks git:(function-name) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkAccumulatedContext/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkAccumulatedContext/Zap.Check-12             11006792           117 ns/op           0 B/op           0 allocs/op
BenchmarkAccumulatedContext/Zap.CheckSampled-12      29545087            38.7 ns/op           0 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    2.718s
➜  benchmarks git:(function-name) ✗ go test -cpuprofile=cpu.pprof -memprofile=mem.pprof -benchmem -run="^$$" -bench=BenchmarkAddingFields/Zap.Check
goos: darwin
goarch: amd64
pkg: go.uber.org/zap/benchmarks
BenchmarkAddingFields/Zap.Check-12              1000000          1011 ns/op         772 B/op           5 allocs/op
BenchmarkAddingFields/Zap.CheckSampled-12       7839132           147 ns/op          86 B/op           0 allocs/op
PASS
ok      go.uber.org/zap/benchmarks    2.474s
```